src/sage/groups/additive_abelian/additive_abelian_wrapper.py

```diff
@@ -566,3 +616,188 @@ def _rec(j, k, c):
+    r"""
+    Given a generating set of some finite abelian group
+    (additively written), compute and return a basis of
+    the group.
+
+    .. NOTE::
+
+        A *basis* of a finite abelian group is a generating
+        set `\{g_1,...,g_n\}` such that each element of the
+        group can be written as a unique linear combination
+        `\alpha_1 g_1 + ... + \alpha_n g_n` with each
+        `\alpha_i \in \{0,...,\mathrm{ord}(g_i)-1\}`.
+
+    ALGORITHM: [Suth2007]_, Algorithm 9.1 & Remark 9.1
+
+    EXAMPLES::
+
+        sage: from sage.groups.additive_abelian.additive_abelian_wrapper import basis_from_generators
+        sage: E = EllipticCurve(GF(31337^6,'a'), j=37)
+        sage: E.order()
+        946988065073788930380545280
+        sage: (R,S), (ordR,ordS) = basis_from_generators(E.gens())
+        sage: ordR, ordS
+        (313157428926517503432720, 3024)
+        sage: R.order() == ordR
+        True
+        sage: S.order() == ordS
+        True
+        sage: ordR * ordS == E.order()
+        True
+        sage: R.weil_pairing(S, ordR).multiplicative_order() == ordS
+        True
+        sage: E.abelian_group().invariants()
+        (3024, 313157428926517503432720)
+    """
+    if ords is None:
+        ords = [g.order() for g in gens]
+    if not gens:
+        return [], []
+
+    from sage.arith.functions import lcm
+    lam = lcm(ords)
+    ps = sorted(lam.prime_factors(), key=lam.valuation)
+
+    gammas, ms = [], []
+    for p in ps:
+        pgens = [(o.prime_to_m_part(p) * g, o.p_primary_part(p)) for g,o in zip(gens,ords) if not o%p]
+        assert pgens
+        pgens.sort(key=lambda tup: tup[1])
+
+        alpha, ord_alpha = pgens.pop()
+        vals = [ord_alpha.valuation(p)]
+        alphas = [alpha]
+        del ord_alpha, alpha
+
+        while pgens:
+            beta, ord_beta = pgens.pop()
+            try:
+                dlog = _discrete_log_pgroup(p, vals, alphas, beta)
+            except TypeError:
+                pass
+            else:
+                continue
+
+            # step 4
+            val_beta = ord_beta.valuation(p)
+            beta_q = beta
+            for v in range(1, val_beta):
+                beta_q *= p
+#                assert beta_q == beta * p**v
+                try:
+                    e = _discrete_log_pgroup(p, vals, alphas, -beta_q)
+                except TypeError:
+                    continue
+                else:
+                    _expand_basis_pgroup(p, alphas, vals, beta, val_beta, list(e) + [p**v])
+#                    assert all(a.order() == p**v for a,v in zip(alphas, vals))
+                    break
+            else:
+                alphas.append(beta)
+                vals.append(val_beta)
+
+        for i,(v,a) in enumerate(reversed(sorted(zip(vals, alphas)))):
+            if i < len(gammas):
+                gammas[i] += a
+                ms[i] *= p**v
+            else:
+                gammas.append(a)
+                ms.append(p**v)
```

Comment:
```suggestion
        for i, (v, a) in enumerate(sorted(zip(vals, alphas), reverse=True)):
            if i < len(gammas):
                gammas[i] += a
                ms[i] *= p ** v
            else:
                gammas.append(a)
                ms.append(p ** v)
```
PEP8 and a simplification.